### DISCUSSION TITLE
Palettized Color completely broken?

### DISCUSSION THREAD

**Discussion by TheRealQuantam** [python-pillow/Pillow#6388] [Q&A] [answered]
I'm encountering several severe issues with palettized color in Pillow 9.1.1 on Python 3.7 64-bit. I'm posting this in discussions in case I am doing something wrong, but if I'm not, I'll elevate these to issues. Per request I'm splitting off each issue into a separate thread.

Image.save to GIF does not use Image.palette, but requires you to specify the palette with the palette parameter to Image.save. If you do not specify the palette in Image.save it treats the image as greyscale. The situation appears to be worse with PCX, as I don't see ANY way to make it use the palette with it.

Partial code:
```python
img = Image.fromarray(bitmap, "P")
img.palette = img_pal
img.save("saved as greyscale not palette.gif", transparency = -1)
img.save("saved as greyscale not palette.png", palette = img.palette)
img.save("saved as greyscale not palette.pcx", palette = img.palette)
```

[bad_pillow.zip](https://github.com/python-pillow/Pillow/files/8954031/bad_pillow.zip)

Correct version:
![manually remapped palette](https://user-images.githubusercontent.com/54358053/174932980-ee0b10d6-ce1a-431d-90f0-08e4e41c0efc.gif)

Saved as greyscale when they should have been saved with palette:
![saved as greyscale not palette](https://user-images.githubusercontent.com/54358053/174932983-29fd4331-95c1-4031-ad6f-33f50625beea.gif)
![converted as greyscale not palette](https://user-images.githubusercontent.com/54358053/174936840-683b71f1-4315-4d9e-9c64-4ac1ba8fd60c.png)

**Comment by pinobatch** [comment]
I'd be interested to see a minimal reproducible example.

**Comment by radarhere** [comment]
Hi. Thanks for starting to provide examples. I think it would be simpler if you created three different threads for this, each with example code and images.

**Comment by TheRealQuantam** [comment]
Replacing "img.palette = img_pal" with "img.putpalette(img_pal)" has partially solved the problem. PNG, PCX, and GIF no longer needs the palette explicitly specified (the color remapping issue in the other thread remains for GIF, though). BUT, for PCX the palette is all messed up (this occurs regardless of whether the palette is specified at save). Though I'm not as concerned about this, as it was just another palettized format I used to see if the same thing would happen as with GIF.

PCX converted to PNG:
![saved as greyscale not palette pcx](https://user-images.githubusercontent.com/54358053/174954363-8d264bf4-4564-4619-9106-8c000e4f8c70.png)

**Reply by radarhere** [reply, marked as answer]
I've created PR #6391 to fix this.

**Reply by radarhere** [reply]
The PR has been merged, and will be part of the next Pillow release, due out on July 1
